[IsDeprecatedWeakRefSmartPointerException] Make PageLoadState and related classes RefCounted or equivalent

basuke opened the pull request:
#### 0eadc6960dc578b3e8c3cc5a7bbcda543ccd0876
<pre>
[IsDeprecatedWeakRefSmartPointerException] Make PageLoadState and related classes RefCounted or equivalent
<a href="https://bugs.webkit.org/show_bug.cgi?id=280282">https://bugs.webkit.org/show_bug.cgi?id=280282</a>

Reviewed by Chris Dumez.

Those classes are CanMakeWeakPtr so that need to be lifecycle manageable.

* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageStateClient):
* Source/WebKit/UIProcess/API/C/mac/WKPagePrivateMac.mm:
(-[WKObservablePageState initWithPage:]):
* Source/WebKit/UIProcess/API/Cocoa/PageLoadStateObserver.h:
* Source/WebKit/UIProcess/API/Cocoa/WKBrowsingContextController.mm:
(-[WKBrowsingContextController _initWithPageRef:]):
(WebKit::PageLoadStateObserver::ref const):
(WebKit::PageLoadStateObserver::deref const):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _initializeWithConfiguration:]):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewConstructed):
(webkitWebViewDispose):
* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::ref const):
(WebKit::NavigationState::deref const):
* Source/WebKit/UIProcess/PageLoadState.cpp:
(WebKit::PageLoadState::callObserverCallback):
* Source/WebKit/UIProcess/PageLoadState.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setPageLoadStateObserver):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: <a href="https://commits.webkit.org/284380@main">https://commits.webkit.org/284380@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e3e7f0a9dbe5d0387c4a2cb0cde470eebb9d27ab

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69115 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48515 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21787 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73196 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20273 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56316 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20122 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55004 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13452 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72181 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44283 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59667 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35481 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40951 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18647 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62894 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17442 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74907 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13097 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16679 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62658 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13135 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59750 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62557 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15362 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10559 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4169 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44319 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45392 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46588 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45134 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/Cocoa/NavigationState.h:
```diff
@@ -64,13 +55,20 @@ class SecurityOriginData;
+class WebPageLoadTiming;
 
-class NavigationState final : public PageLoadState::Observer {
+class NavigationState final : public RefCounted<NavigationState>, public PageLoadState::Observer {
```
Aren't we able to forward the ref-counting to the WKWebView instead? It seems to own the NavigationState.

basuke replied:
Trying.

cdumez commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -243,15 +243,22 @@ enum {
 static GParamSpec* sObjProperties[N_PROPERTIES] = { nullptr, };
 
-class PageLoadStateObserver final : public PageLoadState::Observer {
+class PageLoadStateObserver final : public RefCounted<PageLoadStateObserver> public PageLoadState::Observer {
```
Missing comma?

basuke replied:
Yes :( Lost 1 hour.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/PageLoadStateObserver.h:
```diff
@@ -29,21 +29,28 @@
 namespace WebKit {
 
-class PageLoadStateObserver : public PageLoadState::Observer {
+class PageLoadStateObserver : public RefCounted<PageLoadStateObserver>, public PageLoadState::Observer {
```
Can't we forward the ref counting to the owner (WKBrowsingContextController) ?

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -410,7 +410,7 @@ - (void)_initializeWithConfiguration:(WKWebViewConfiguration *)configuration
     _page->setApplicationNameForDesktopUserAgent(configuration._applicationNameForDesktopUserAgent);
 
-    _navigationState = makeUnique<WebKit::NavigationState>(self);
+    _navigationState = WebKit::NavigationState::create(self);
```
You're introducing a reference cycle and thus a leak here. WKWebView needs to keep using `makeUnique<WebKit::NavigationState>(self);`. Otherwise, WKWebView refs NavigationState and refine the NavigationState refs the WKWebView again (due to ref counting forwarding).

basuke replied:
Ah. Thanks! I should be careful when forwarding ref/deref to other object.

cdumez commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -243,15 +243,22 @@ enum {
+class PageLoadStateObserver final : public RefCounted<PageLoadStateObserver>, public PageLoadState::Observer {
     WTF_MAKE_TZONE_ALLOCATED_INLINE(PageLoadStateObserver);
 public:
+    static RefPtr<PageLoadStateObserver>create(WebKitWebView* webView)
```
Missing space before `create(`

cdumez commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -2440,14 +2440,22 @@ void WKPageSetPageNavigationClient(WKPageRef pageRef, const WKPageNavigationClie
+class StateClient final : public RefCounted<StateClient>, public API::Client<WKPageStateClientBase>, public PageLoadState::Observer {
     WTF_MAKE_TZONE_ALLOCATED_INLINE(StateClient);
 public:
+    static RefPtr<StateClient> create(const WKPageStateClientBase* client)
```
Should return a `Ref<>` since it cannot return nullptr.

cdumez commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -2440,14 +2440,22 @@ void WKPageSetPageNavigationClient(WKPageRef pageRef, const WKPageNavigationClie
 public:
+    static RefPtr<StateClient> create(const WKPageStateClientBase* client)
+    {
+        return adoptRef(new StateClient(client));
```
`return adoptRef(*new StateClient(client));`

cdumez commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -243,15 +243,22 @@ enum {
+class PageLoadStateObserver final : public RefCounted<PageLoadStateObserver>, public PageLoadState::Observer {
     WTF_MAKE_TZONE_ALLOCATED_INLINE(PageLoadStateObserver);
 public:
+    static RefPtr<PageLoadStateObserver> create(WebKitWebView* webView)
```
Should return a `Ref<>`

cdumez commented on Source/WebKit/UIProcess/Cocoa/NavigationState.h:
```diff
@@ -64,13 +54,17 @@ class SecurityOriginData;
     WTF_MAKE_TZONE_ALLOCATED(NavigationState);
 public:
-    explicit NavigationState(WKWebView *);
+    explicit NavigationState(WKWebView*);
```
The previous style was correct, since this is an Objective C type.

cdumez reviewed (approved):
LGTM with changes.

basuke committed:
[IsDeprecatedWeakRefSmartPointerException] Make PageLoadState and related classes RefCounted or equivalent
https://bugs.webkit.org/show_bug.cgi?id=280282

Reviewed by Chris Dumez.

Those classes are CanMakeWeakPtr so that need to be lifecycle manageable.

* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageStateClient):
* Source/WebKit/UIProcess/API/C/mac/WKPagePrivateMac.mm:
(-[WKObservablePageState initWithPage:]):
* Source/WebKit/UIProcess/API/Cocoa/PageLoadStateObserver.h:
* Source/WebKit/UIProcess/API/Cocoa/WKBrowsingContextController.mm:
(-[WKBrowsingContextController _initWithPageRef:]):
(WebKit::PageLoadStateObserver::ref const):
(WebKit::PageLoadStateObserver::deref const):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _initializeWithConfiguration:]):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewConstructed):
(webkitWebViewDispose):
* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::ref const):
(WebKit::NavigationState::deref const):
* Source/WebKit/UIProcess/PageLoadState.cpp:
(WebKit::PageLoadState::callObserverCallback):
* Source/WebKit/UIProcess/PageLoadState.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setPageLoadStateObserver):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: https://commits.webkit.org/284380@main
